File: Algo/Leetcode/179LargestNum.cpp
class Solution {
public:
    string largestNumber(vector<int>& nums) {
        sort(nums.begin(), nums.end(), comp);
        string s = "";
        for(auto n:nums) s += to_string(n);
        int start = 0;
        while(start < s.length() && s[start]=='0') start++;
        if(start == s.length()) return "0";
        return s.substr(start);
    }

    static bool comp(int a, int b){
        string m = to_string(a) + to_string(b);
        string n = to_string(b) + to_string(a);
        return m>n;
    }
};

File: Algo/Leetcode/179LargestNum.py
class Solution(object):
    def largestNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: str
        """
        def comp(k1, k2):
            s1 = int(str(k1) + str(k2))
            s2 = int(str(k2) + str(k1))
            if s1 < s2:
                return 1
            else:
                return -1

        nums = sorted(nums, key=functools.cmp_to_key(comp))

        # consider all 0 case
        if nums[0] == nums[-1] == 0:
            return "0"

        return ''.join([str(n) for n in nums])